taichi/program/async_utils.cpp

```diff
@@ -162,19 +175,72 @@ TaskMeta *get_task_meta(IRBank *ir_bank, const TaskLaunchRecord &t) {
     if (auto clear_list = stmt->cast<ClearListStmt>()) {
       meta.output_states.emplace(clear_list->snode, AsyncState::Type::list);
     }
-    // TODO: handle SNodeOpStmt etc.
     return false;
   });
 
-  // We are being conservative here: if there are any non-element-wise
-  // accesses (e.g., a = x[i + 1]), we don't treat it as completely
-  // overwriting the value state (e.g., for i in x: x[i] = 0).
-  for (auto &state : meta.output_states) {
-    if (state.type == AsyncState::Type::value) {
-      if (meta.element_wise.find(state.snode) == meta.element_wise.end()) {
-        meta.input_states.insert(state);
+  std::unordered_set<SNode *> kernel_forces_no_activate(
+      t.kernel->no_activate.begin(), t.kernel->no_activate.end());
+
+  std::unordered_set<SNode *> mask_state_inserted;
+  auto insert_mask_states_for_activation = [&](SNode *s) {
+    while (s) {
+      if (kernel_forces_no_activate.count(s) > 0) {
+        break;
+      }
+      if (mask_state_inserted.count(s) > 0) {
+        // already handled by other activations
+        break;
+      }
+      mask_state_inserted.insert(s);
+
+      // Do not record dense SNodes' mask states.
+      if (s->type != SNodeType::dense && s->type != SNodeType::place &&
+          s->type != SNodeType::root) {
+        meta.input_states.emplace(s, AsyncState::Type::mask);
+        meta.output_states.emplace(s, AsyncState::Type::mask);
+        if (is_gc_able(s->type)) {
+          meta.input_states.emplace(s, AsyncState::Type::allocator);
+          meta.output_states.emplace(s, AsyncState::Type::allocator);
+        }
       }
+      s = s->parent;
     }
+  };
+
+  for (auto &snode : activates) {
+    insert_mask_states_for_activation(snode);
+  }
+  for (auto &snode : deactivates) {
+    insert_mask_states_for_activation(snode);
+  }
+
+  auto insert_value_states_for_deactivation = [&](SNode *snode) {
+    // Insert output value states for all successors of snode.
+    // Input value states will be inserted later if it's not
+    // element-wise written.
+    std::queue<SNode *> to_insert;
+    to_insert.push(snode);
+    while (!to_insert.empty()) {
+      auto *s = to_insert.front();
+      to_insert.pop();
+      if (kernel_forces_no_activate.count(s) > 0) {
+        continue;
+      }
+      if (s->type == SNodeType::place) {
+        meta.output_states.emplace(s, AsyncState::Type::value);
```

Comment:
Originally, I was going to ask if we should also include all `b`'s descendants' mask states. For example:

```py
a = ti.root.pointer(...)
b = a.dense(...)
c = b.pointer(...)
c.place(d)

ti.deactivate(b, I)
```

My reasoning: Upon the next time we run listgen, because `b` is deactivated at index `I`, all `b`'s descendant cells covered by `I`, including `c` (a `dense` SNode) will be skipped. So from this perspective, `deactivate` actually changes its children's mask states as well.

---

However I think what I've said above is probably already handled by the state dependencies between two `listgen` tasks. For example, if we have `listgen A` and `listgen B`, where `A` is a parent of `B`, we insert `(A, list)` to `B`'s listgen task meta:

https://github.com/taichi-dev/taichi/blob/2f809aa6b311a3019308c8612ff3c4fdefe07854/taichi/program/async_utils.cpp#L183

But do we also need to insert `(B, list)` into `A`'s meta?

Reply:
I think we need to also include all `b`'s descendants' mask states in gc tasks. We don't want the following gc and listen tasks to be swapped:
```
gc(b) (input&output (b, allocator), (d, value))
clear_list(c) (output (c, list))
listgen(c) (input (b, list), (c, list), (c, mask), output (c, list))
```

Reply:
(I think we should include these states in the meta of gc(b) (suppose b is gc-able): `input&output (b, mask), (b, allocator), (c, mask), (c, allocator), (d, value)`)

Reply:
Hmm, i feel like it's fine to just include `(b, allocator), (b, mask)` inside `gc(b)` task itself? IIUC, here's how it would look like

```
user_task_that_deactivates_b (input&output (b, allocator), (d, value))  <-- missing (b, mask)?
gc(b) ((b, allocator), (b, mask))

# listgen always starts from the root, so b's listgen will always show up in c's.
clear_list(b) (output (b, list))
listgen(b) (input (a, list), (b, list), (b, mask), output (b, list))
clear_list(c) (output (c, list))
listgen(c) (input (b, list), (c, list), (c, mask), output (c, list))
```

Reply:
I see. But what about this:
```
user_task_that_deactivates_b (input&output (b, allocator),  (b, mask), (d, value))
gc(b) ((b, allocator), (b, mask))
user_task_that_deactivates_c (input&output (c, allocator),  (c, mask), (d, value))
gc(c) ((c, allocator), (c, mask))
user_task_that_activates_a (can this happen?) (input&output (a, allocator),  (a, mask))
```
Is it fine to reorder these tasks like this?
```
user_task_that_activates_a (can this happen?) (input&output (a, allocator),  (a, mask))
user_task_that_deactivates_b (input&output (b, allocator),  (b, mask), (d, value))
user_task_that_deactivates_c (input&output (c, allocator),  (c, mask), (d, value))
gc(c) ((c, allocator), (c, mask))
gc(b) ((b, allocator), (b, mask))
```